use FastIntSet for bitmaps

RaduBerinde opened the pull request:
This change also helps readability because a lot of direct bit operations are
replaced with properly named function calls.

Depends on https://github.com/cockroachdb/cockroach/pull/20095

I can add the "range compression" to `FastIntSet.String` if you feel strongly about it.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/petermattis/opttoy/23)
<!-- Reviewable:end -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> though I'm curious whether `FastIntSet` is the right data structure here. More specifically, I'm wondering if `intsets.Sparse` is the right data structure. Won't the bitmaps used during optimization generally be dense?

I liked the range display of `bitmap.String`, but don't feel strongly about it.

---

Review status: 0 of 27 files reviewed at latest revision, 1 unresolved discussion.

---

*[v3/bitmap.go, line 8 at r1](https://reviewable.io:443/reviews/petermattis/opttoy/23#-Kz5FmJTSmllNQFPDVfk:-Kz5FmJTSmllNQFPDVfl:b-35zncs) ([raw file](https://github.com/petermattis/opttoy/blob/d63e96d98dad8b23a7ed4cead809db1eb3558516/v3/bitmap.go#L8)):*
> ```Go
> 
> // Bitmap used for columns.
> type bitmap = util.FastIntSet
> ```

Type aliases!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/petermattis/opttoy/23#-:-Kz5G0rmWy_5p7NpgRo4:b-jz525s)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
`Sparse` works best on dense sets (despite the name). It's a linked list of 256-bit bitmaps; each block can start at an arbitrary index (as long as it's a multiple of 256)

---

Review status: 0 of 27 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/petermattis/opttoy/23#-:-Kz5HzPySqDfKxka3s8N:blyj21i)*
<!-- Sent from Reviewable.io -->

petermattis commented:
The other representation I was thinking of was `[]uint32` where the slices are allocated to the desired size.I would guess most queries will use <256 bits. There is a bit of wastage in the `Sparse` representation.

We can revisit this later. I think this change is a good step forward because you replaced all the implicit bitmap operations with methods. Switching to another representation in the future will now be easy.

RaduBerinde commented:
I like the range display too, added it to FastIntSet and updated this PR.

Yeah, we can tweak the format if necessary. My thinking is that in the grand scheme of things, running a query where things have > 64 columns is going to be slow enough anyway for those few wasted words to not matter. And it's more convenient not having to size the set upfront.

---

Review status: 0 of 27 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/petermattis/opttoy/23#-:-Kz5NnvLCCLjV6J6k46C:bbunft3)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Roger. :shipit: 

---

Review status: 0 of 27 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/petermattis/opttoy/23#-:-Kz5_oQges_7VUfCdWO8:b-7xgcyl)*
<!-- Sent from Reviewable.io -->
